spec/models/address_spec.rb

```diff
@@ -253,5 +253,115 @@ def addr_details(addr, expected_msg)
 
     end
 
+
+    describe '#self.geocode_all_needed(sleep_between: 0.5, num_per_batch: 50)' do
+
+      let(:a_company) { create(:company, num_addresses: 0) }
+
+      let(:norbotten_region) { create(:region, name: 'Norrbotten') }
+      let(:overtornea_kommun) { create(:kommun, name: 'Övertorneå') }
+
+      # These are real addresses in  Övertorneå Municipality in Norrbotten County:
+
+      let(:valid_address1) { addr1 = create(:address,
+                                    street_address: 'Matarengivägen 24',
+                                    post_code: '957 31',
+                                    city: 'Övertorneå',
+                                    kommun: overtornea_kommun,
+                                    region: norbotten_region,
+                                    addressable: a_company )
+                            addr1.validate
+                            addr1
+       }
+
+      let(:valid_address2) { addr2 = create(:address,
+                                    street_address: 'Skolvägen 12',
+                                    post_code: '957 31',
+                                    city: 'Övertorneå',
+                                    kommun: overtornea_kommun,
+                                    region: norbotten_region,
+                                    addressable: a_company )
+                            addr2.validate
+                            addr2
+       }
+
+      let(:valid_address3) { addr3 = create(:address,
+                                            street_address: 'Matarengivägen 30',
+                                            post_code: '957 31',
+                                            city: 'Övertorneå',
+                                            kommun: overtornea_kommun,
+                                            region: norbotten_region,
+                                            addressable: a_company )
+                              addr3.validate
+                              addr3
+       }
+
+      it 'nothing geocoded if all have latitude and longitude' do
+        valid_address1
+        valid_address2
+        valid_address3
+
+        need_geocoding = Address.not_geocoded
+        needed_geocoding = need_geocoding.count
+
+        Address.geocode_all_needed
+
+        after_run_need_geocoding = Address.not_geocoded.count
+
+        expect(needed_geocoding).to eq 0
+        expect(after_run_need_geocoding).to eq 0
+      end
+
+
+      it 'will geocode 1 that needs it' do
+
+        valid_address1
+        valid_address2
+        valid_address3
+
+        query = <<-SQL
+          UPDATE addresses SET latitude=NULL, longitude=NULL  
+           WHERE street_address = 'Matarengivägen 24'
+        SQL
```

Comment:
It's not the geocoder gem, it's us.

When a single company is _fetched_ from the db, a check is done to see if it (the addresses for it) needs to be geocoded.  (This happens during `set_company` in the CompaniesController before the `show` action)

After an Address is _validated_, it's geocoded.  (This happens with  `after_validation :geocode_best_possible,   :if => lambda { |obj| obj.changed? } ` in the Address model)  This is the typical pattern when using Geocoder.

So we have an Address geocoded when we read it in (via a Company) and when we save it.

It seemed easiest to me to just write straight SQL to set up the circumstances.  I couldn't think of a straightforward way to get around the geocoding other than that.  


As always, I'm delighted to change my code for one of your excellent suggestions. :-)